File: OneSignalSDK/onesignal/src/main/java/com/onesignal/NotificationGenerationJob.java
/**
 * Modified MIT License
 *
 * Copyright 2017 OneSignal
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.onesignal;


import android.content.Context;
import android.net.Uri;

import org.json.JSONObject;

import java.security.SecureRandom;

class NotificationGenerationJob {
   Context context;
   JSONObject jsonPayload;
   boolean restoring;
   
   boolean showAsAlert;
   
   Long shownTimeStamp;
   
   CharSequence overriddenBodyFromExtender;
   CharSequence overriddenTitleFromExtender;
   Uri overriddenSound;
   Integer overriddenFlags;
   Integer orgFlags;
   Uri orgSound;
   
   NotificationGenerationJob(Context context) {
      this.context = context;
   }
   
   NotificationExtenderService.OverrideSettings overrideSettings;
   
   CharSequence getTitle() {
      if (overriddenTitleFromExtender != null)
         return overriddenTitleFromExtender;
      return jsonPayload.optString("title", null);
   }
   
   CharSequence getBody() {
      if (overriddenBodyFromExtender != null)
         return overriddenBodyFromExtender;
      return jsonPayload.optString("alert", null);
   }
   
   Integer getAndroidId() {
      if (overrideSettings == null)
         overrideSettings = new NotificationExtenderService.OverrideSettings();
      if (overrideSettings.androidNotificationId == null)
         overrideSettings.androidNotificationId = new SecureRandom().nextInt();
      
      return overrideSettings.androidNotificationId;
   }

   int getAndroidIdWithoutCreate() {
      if (overrideSettings == null || overrideSettings.androidNotificationId == null)
         return -1;

      return overrideSettings.androidNotificationId;
   }

   void setAndroidIdWithOutOverriding(Integer id) {
      if (id == null)
         return;

      if (overrideSettings != null && overrideSettings.androidNotificationId != null)
         return;

      if (overrideSettings == null)
         overrideSettings = new NotificationExtenderService.OverrideSettings();
      overrideSettings.androidNotificationId = id;
   }

   boolean hasExtender() {
      return overrideSettings != null && overrideSettings.extender != null;
   }
}

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/internal/notification/work/NotificationGenerationJob.kt
package com.onesignal.onesignal.internal.notification.work

import android.content.Context
import android.net.Uri
import com.onesignal.onesignal.internal.common.time.ITime
import com.onesignal.onesignal.internal.notification.Notification
import com.onesignal.onesignal.internal.notification.NotificationHelper
import org.json.JSONObject
import java.security.SecureRandom

class NotificationGenerationJob {
    var notification: Notification = Notification()
    var context: Context
    var jsonPayload: JSONObject? = null
    var isRestoring = false
    var shownTimeStamp: Long? = null
    var overriddenBodyFromExtender: CharSequence? = null
    var overriddenTitleFromExtender: CharSequence? = null
    var overriddenSound: Uri? = null
    var overriddenFlags: Int? = null
    var orgFlags: Int? = null
    var orgSound: Uri? = null

    internal constructor(context: Context) {
        this.context = context
    }

    internal constructor(context: Context, jsonPayload: JSONObject?, time: ITime) : this(
        context, Notification(jsonPayload!!, time), jsonPayload
    )

    internal constructor(context: Context, notification: Notification, jsonPayload: JSONObject?) {
        this.context = context
        this.jsonPayload = jsonPayload
        this.notification = notification
    }

    /**
     * Get the notification title from the payload
     */
    val title: CharSequence
        get() = overriddenTitleFromExtender ?: notification!!.title as CharSequence

    /**
     * Get the notification body from the payload
     */
    val body: CharSequence
        get() = overriddenBodyFromExtender ?: notification!!.body as CharSequence

    /**
     * Get the notification additional data json from the payload
     */
    val additionalData: JSONObject
        get() = notification!!.additionalData ?: JSONObject()

    /**
     * If androidNotificationId is -1 then the notification is a silent one
     */
    val isNotificationToDisplay: Boolean
        get() = androidIdWithoutCreate != -1

    fun hasExtender(): Boolean {
        return notification!!.notificationExtender != null
    }

    val apiNotificationId: String
        get() = NotificationHelper.getNotificationIdFromFCMJson(jsonPayload) ?: ""

    val androidIdWithoutCreate: Int
        get() = if (!notification!!.hasNotificationId()) -1 else notification!!.androidNotificationId

    val androidId: Int
        get() {
            if (!notification!!.hasNotificationId())
                notification!!.androidNotificationId = SecureRandom().nextInt()

            return notification!!.androidNotificationId
        }

    fun setAndroidIdWithoutOverriding(id: Int?) {
        if (id == null) return
        if (notification!!.hasNotificationId()) return
        notification!!.androidNotificationId = id
    }

    override fun toString(): String {
        return "OSNotificationGenerationJob{" +
                "jsonPayload=" + jsonPayload +
                ", isRestoring=" + isRestoring +
                ", shownTimeStamp=" + shownTimeStamp +
                ", overriddenBodyFromExtender=" + overriddenBodyFromExtender +
                ", overriddenTitleFromExtender=" + overriddenTitleFromExtender +
                ", overriddenSound=" + overriddenSound +
                ", overriddenFlags=" + overriddenFlags +
                ", orgFlags=" + orgFlags +
                ", orgSound=" + orgSound +
                ", notification=" + notification +
                '}'
    }
}